File: frontend/src/types/ResourcePlanning.ts
import type { TeamResourcePlanningRow } from './Team';

export type ResourcePlanningGranularity = 'day' | 'week' | 'month';

export type ResourceStatusColor = 'green' | 'blue' | 'orange' | 'red' | 'grey';

export interface ResourceAllocationBlock {
  project_id: string;
  tool_number: string;
  customer_name: string;
  milestone_name: string | null;
  hours: number;
  status_color: ResourceStatusColor;
  complexity?: string | null;
}

export interface ResourcePlanningPeriod {
  key: string;
  label: string;
  start_date: string;
  end_date: string;
}

export interface ResourcePlanningCell {
  period_key: string;
  capacity_hours: number;
  allocated_hours: number;
  remaining_hours: number;
  status_color: ResourceStatusColor;
  blocks: ResourceAllocationBlock[];
}

export interface ResourcePlanningDesignerRow {
  user_id: string;
  designer_name: string;
  team_name: string | null;
  availability_status: string;
  skill_level?: string | null;
  capacity_hours: number;
  allocated_hours: number;
  remaining_hours: number;
  cells: ResourcePlanningCell[];
}

export interface UnassignedProjectBlock {
  project_id: string;
  tool_number: string;
  customer_name: string;
  quoted_hours: number;
  remaining_hours: number;
  due_date: string | null;
  milestone_name: string | null;
  complexity?: string | null;
}

export interface ResourcePlanningGrid {
  granularity: ResourcePlanningGranularity;
  start_date: string;
  end_date: string;
  periods: ResourcePlanningPeriod[];
  designers: ResourcePlanningDesignerRow[];
  unassigned_projects: UnassignedProjectBlock[];
  team_summary: TeamResourcePlanningRow[];
}

/** IT readiness — read-only projections over IT Operations (Wave 4). */

export type ITGapSeverity = 'error' | 'warning' | 'success' | 'info';

export interface ITGapAlert {
  severity: ITGapSeverity;
  code: string;
  message: string;
  count: number;
}

export interface ITGapUser {
  user_id: string;
  full_name: string;
  email: string | null;
}

export interface ITGapUserMissingLicenses extends ITGapUser {
  missing: { software_id: string; software_name: string | null }[];
}

export interface ITLicenseDemandRow {
  software_id: string;
  software_name: string | null;
  required_headcount: number;
  unmet_headcount: number;
  seat_count: number;
  assigned_seats: number;
  available_seats: number;
  is_floating: boolean;
  peak_estimate: number;
  seat_shortfall: number;
  has_pool: boolean;
}

export interface ITLicensePoolRow {
  pool_id: string;
  software_id: string;
  software_name: string | null;
  license_type: string | null;
  is_floating: boolean;
  seat_count: number;
  assigned_seats: number;
  available_seats: number;
  expiry_date: string | null;
  is_expired: boolean;
  expires_soon: boolean;
}

export interface ResourceITGaps {
  on_date: string;
  team_id: string | null;
  headcount: number;
  users_without_computer: ITGapUser[];
  users_missing_licenses: ITGapUserMissingLicenses[];
  license_demand: ITLicenseDemandRow[];
  oversubscribed_software: ITLicenseDemandRow[];
  software_without_pool: ITLicenseDemandRow[];
  expiring_pools: ITLicensePoolRow[];
  expired_pools: ITLicensePoolRow[];
  spare_computers: number;
  total_computers: number;
  alerts: ITGapAlert[];
}

export interface ResourceITMatrixRow {
  user_id: string;
  full_name: string;
  email: string | null;
  team_name: string | null;
  has_computer: boolean;
  computer_name: string | null;
  asset_number: string | null;
  required_software_count: number;
  licensed_software_count: number;
  missing_software: string[];
  is_compliant: boolean;
  is_ready: boolean;
}

export interface ResourceITMatrix {
  from_date: string;
  to_date: string;
  team_id: string | null;
  headcount: number;
  ready_count: number;
  not_ready_count: number;
  rows: ResourceITMatrixRow[];
}

File: app/schemas/resource_planning.py
"""Resource planning grid — designers × time periods."""

from __future__ import annotations

from datetime import date, timedelta
from decimal import Decimal
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.reports import TeamResourcePlanningRow


class ResourcePlanningGranularity(str, Enum):
    day = "day"
    week = "week"
    month = "month"


class ResourceAllocationBlock(BaseModel):
    project_id: UUID
    tool_number: str
    customer_name: str
    milestone_name: str | None = None
    hours: Decimal
    status_color: str
    complexity: str | None = None


class ResourcePlanningPeriod(BaseModel):
    key: str
    label: str
    start_date: date
    end_date: date


class ResourcePlanningCell(BaseModel):
    period_key: str
    capacity_hours: Decimal
    allocated_hours: Decimal
    remaining_hours: Decimal
    status_color: str
    blocks: list[ResourceAllocationBlock] = Field(default_factory=list)


class ResourcePlanningDesignerRow(BaseModel):
    user_id: UUID
    designer_name: str
    team_name: str | None = None
    availability_status: str
    skill_level: str | None = None
    capacity_hours: Decimal
    allocated_hours: Decimal
    remaining_hours: Decimal
    cells: list[ResourcePlanningCell] = Field(default_factory=list)


class UnassignedProjectBlock(BaseModel):
    project_id: UUID
    tool_number: str
    customer_name: str
    quoted_hours: Decimal
    remaining_hours: Decimal
    due_date: date | None = None
    milestone_name: str | None = None
    complexity: str | None = None


class ResourcePlanningGrid(BaseModel):
    granularity: ResourcePlanningGranularity
    start_date: date
    end_date: date
    periods: list[ResourcePlanningPeriod] = Field(default_factory=list)
    designers: list[ResourcePlanningDesignerRow] = Field(default_factory=list)
    unassigned_projects: list[UnassignedProjectBlock] = Field(default_factory=list)
    team_summary: list[TeamResourcePlanningRow] = Field(default_factory=list)


class ResourcePlanningAssignRequest(BaseModel):
    project_id: UUID
    designer_id: UUID | None = None


# --- IT readiness (Connected ProTrack Wave 4) ------------------------------
# Read-only projections over the IT Operations source of truth. Resource
# Planning owns no computer or software master data of its own.


class ITGapAlert(BaseModel):
    severity: str
    code: str
    message: str
    count: int = 0


class ITGapUser(BaseModel):
    user_id: UUID
    full_name: str
    email: str | None = None


class ITGapMissingSoftware(BaseModel):
    software_id: UUID
    software_name: str | None = None


class ITGapUserMissingLicenses(ITGapUser):
    missing: list[ITGapMissingSoftware] = Field(default_factory=list)


class ITLicenseDemandRow(BaseModel):
    software_id: UUID
    software_name: str | None = None
    required_headcount: int = 0
    unmet_headcount: int = 0
    seat_count: int = 0
    assigned_seats: int = 0
    available_seats: int = 0
    is_floating: bool = False
    peak_estimate: int = 0
    seat_shortfall: int = 0
    has_pool: bool = False


class ITLicensePoolRow(BaseModel):
    pool_id: UUID
    software_id: UUID
    software_name: str | None = None
    license_type: str | None = None
    is_floating: bool = False
    seat_count: int = 0
    assigned_seats: int = 0
    available_seats: int = 0
    expiry_date: date | None = None
    is_expired: bool = False
    expires_soon: bool = False


class ResourceITGaps(BaseModel):
    on_date: date
    team_id: UUID | None = None
    headcount: int = 0
    users_without_computer: list[ITGapUser] = Field(default_factory=list)
    users_missing_licenses: list[ITGapUserMissingLicenses] = Field(default_factory=list)
    license_demand: list[ITLicenseDemandRow] = Field(default_factory=list)
    oversubscribed_software: list[ITLicenseDemandRow] = Field(default_factory=list)
    software_without_pool: list[ITLicenseDemandRow] = Field(default_factory=list)
    expiring_pools: list[ITLicensePoolRow] = Field(default_factory=list)
    expired_pools: list[ITLicensePoolRow] = Field(default_factory=list)
    spare_computers: int = 0
    total_computers: int = 0
    alerts: list[ITGapAlert] = Field(default_factory=list)


class ResourceITMatrixRow(BaseModel):
    user_id: UUID
    full_name: str
    email: str | None = None
    team_name: str | None = None
    has_computer: bool = False
    computer_name: str | None = None
    asset_number: str | None = None
    required_software_count: int = 0
    licensed_software_count: int = 0
    missing_software: list[str] = Field(default_factory=list)
    is_compliant: bool = True
    is_ready: bool = False


class ResourceITMatrix(BaseModel):
    from_date: date
    to_date: date
    team_id: UUID | None = None
    headcount: int = 0
    ready_count: int = 0
    not_ready_count: int = 0
    rows: list[ResourceITMatrixRow] = Field(default_factory=list)
